### PR TITLE
ffmpeg

### LINKED ISSUE #10583: VideoWriter ffmpeg pixel format lossy for lossless encodings
##### System information (version)
- OpenCV => master
- Operating System / Platform => Any
- Compiler => Any

##### Objective
Reduce image errors when dealing with the following situations involving VideoWriter with ffmpeg backend
- lossless codecs of RGB data (e.g. PNG, FFV1)
- rawvideo output of RGB data (for applicable container files, e.g. AVI with RGBA fourcc)
- grayscale 16 bit writing (e.g. as simply as PNG encoding)
- RGBA data

##### Detailed description

When using lossless codecs with ffmpeg backend the pixel format selected is lossy by encoding the pixels as yuv420p. This affects the useful lossless encodings such as ffv1 huffyuv and png

ffmpeg -h encoder=ffv1 -v quiet
ffmpeg -h encoder=huffyuv -v quiet
ffmpeg -h encoder=png -v quiet
ffmpeg -h encoder=jpegls -v quiet

FFV1 supports: yuv420p yuva420p yuva422p yuv444p yuva444p yuv440p yuv422p yuv411p yuv410p bgr0 bgra yuv420p16le yuv422p16le yuv444p16le yuv444p9le yuv422p9le yuv420p9le yuv420p10le yuv422p10le yuv444p10le yuv420p12le yuv422p12le yuv444p12le yuva444p16le yuva422p16le yuva420p16le yuva444p10le yuva422p10le yuva420p10le yuva444p9le yuva422p9le yuva420p9le gray16le gray gbrp9le gbrp10le gbrp12le gbrp14le ya8 gray10le gray12le gbrp16le rgb48le

PNG supports: rgb24 rgba rgb48be rgba64be pal8 gray ya8 gray16be ya16be monob

HUFFYUV supports: yuv422p rgb24 bgra

JPEGLS supports: bgr24 rgb24 gray gray16le

Note that even yuv444p can be lossy due to integer arithmetics

##### Steps to reproduce

Easily using VideoWriter and manually generated image files

### PR DESCRIPTION
**Pull request by eruffaldi**
resolves #10583 

### This pullrequest changes
Module: videoio
File: cap_ffmpeg_impl.hpp
Changes: CvVideoWriter_FFMPEG::open
Location: switch of codecs

### CODE REVIEW

**Comment by eruffaldi** [comment]
# Doc Build
trailing error (FIXED)

# Linux Build error
undefined AV_PIX_FMT_BGR0 (depends on ffmpeg version) (FIXED)

# Mac Build error
test_videoio for AVI and fourcc IYUV. This fourcc is considered as rawvideo by ffmpeg (riff.c file { AV_CODEC_ID_RAWVIDEO,     MKTAG('I', 'Y', 'U', 'V') }). But IYUV == I420 and in any case the image will be altered using rawvideo

Hypothesis: different ffmpeg version than Windows build

**Comment by alalek** [comment]
1.

According to this:
https://github.com/opencv/opencv/blame/3.4.0/modules/videoio/src/cap_ffmpeg_impl.hpp#L1923-L1929

there are only two possible values for "input_pix_fmt". Why we need such large switches?

2.

Check in ffmpeg itself (ffplay or extract frames into png) that result file doesn't swap colors (OpenCV's color format is BGR: `red -> Scalar(0, 0, 255)`)

3.

Squash all fixup commits into one. No need to keep them in the history.

**Comment by eruffaldi** [comment]
1. (and 2.) input_pix_fmt only AV_PIX_FMT_BGR24 and AV_PIX_FMT_GRAY8
I acted in this way looking at the original code where the switch was already addressing more than these input formats (e.g. for CODEC_ID_RAWVIDEO):

https://github.com/opencv/opencv/blame/3.4.0/modules/videoio/src/cap_ffmpeg_impl.hpp#L2010-L2013

So I have entered all the input formats that are supported by the specific codec (via ffprobe).

input_pix_fmt should have more values in the case (1) 4 channels  input image, (2) 16-bit single channel input image. This would allow to save alpha channel images and high-depth grayscale images as supported by ffmpeg codecs.

This means two additional changes
1) CvVideoWriter_FFMPEG::writeFrame (easy based on input_pix_fmt)
2) CvVideoWriter_FFMPEG::open (affecting the caller interface)

**Comment by eruffaldi** [comment]
Changed:
- added handling of FOURCC(255,255,255,255) that prints the list of FOURCC for the given output format 
- RAWVIDEO codec given FOURCC=RGBA forces pix_fmt = RGBA
- otherwise RAWVIDEO acts as original code

**Comment by eruffaldi** [comment]
Fixed White space

**Comment by addisonElliott** [comment]
What is stopping this from being merged?

I have an application where I am wanting lossless video I/O. Currently, I am using MJLS codec to have this occur but if I attempt to use rawvideo, it converts to YUV420 and there is some loss in that conversion.

I think the switches containing multiple values are good to have, because it would be nice to expand support for multiple formats in the future.

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 1]
```diff
@@ -1928,6 +1946,13 @@ bool CvVideoWriter_FFMPEG::open( const char * filename, int fourcc,
         input_pix_fmt = AV_PIX_FMT_GRAY8;
     }
 
+    if(fourcc == -1)
+    {
+        fprintf(stderr,"OpenCV: FFMPEG: format %s / %s, codec id %04X\n", fmt->name, fmt->long_name, codec_id);
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: codec_id is always NONE here. Removed

**Pull request merged**